File: Interesectionof2Arrays.java
// Time Complexity :  O(M + N) - shorter string 
// Space Complexity : O(M) - longer string
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No


class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        if(nums1.length > nums2.length) 
            return intersect(nums2, nums1);
        Map<Integer,Integer> map = new HashMap<>();  //O(N)
        for(int i = 0; i < nums2.length; i++){
            if(!map.containsKey(nums2[i]))
                map.put(nums2[i],1);
            else
                map.put(nums2[i], map.get(nums2[i]) + 1);
        }
        List<Integer> res = new ArrayList<>();
        for(int i = 0; i < nums1.length; i++){
            if(map.containsKey(nums1[i]) && map.get(nums1[i]) > 0){
                res.add(nums1[i]);
                map.put(nums1[i], map.get(nums1[i]) - 1);
            }
        }
        int[] result = new int[res.size()];
        
        for(int i = 0; i< res.size();i++){
            result[i] = res.get(i);
        }
        return result;
    }
}
//O(M + N)

File: Interesectionof2arrays.py
# Time Complexity : O(n1+n2).
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : NA

# Approach is to use sorted arrays and two pointers to iterate and find the intersection.
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        n1=len(nums1)
        n2=len(nums2)
        if n1>n2:
            return self.intersect(nums2, nums1)
        p1,p2=0,0
        li=[]
        nums1=sorted(nums1)
        nums2=sorted(nums2)
        while(p1<n1 and p2<n2):
            if(nums1[p1]==nums2[p2]):
                li.append(nums1[p1])
                p1+=1
                p2+=1
            elif(nums1[p1]>nums2[p2]):
                p2+=1
            else:
                p1+=1
        return li
